### PR TITLE
Add structured logging to Koi template

### PR DESCRIPTION
**Pull request by hasarindaKI**

### CODE REVIEW

**Review by AlanCornthwaiteKatalyst — APPROVED** [review]

**Comment by sfnelson** [comment]
@hasarindaKI where did this come from? Can I review what the messages look like somewhere?

File: template.rb [hunk 1]
```diff
@@ -167,6 +168,41 @@ def setup_timezone
...
+  end
+
+  gsub_file("config/environments/production.rb", "::Logger::Formatter", "::StructuredLogging::JsonFormatter")
+  gsub_file("config/environments/production.rb", "ActiveSupport::TaggedLogging.new(logger)",
+            "::StructuredLogging.new(logger)")
+
+  insert_into_file("config/environments/production.rb", before: "config.logger") do
+    "STDOUT.sync = true\n  "
+  end
+  insert_into_file("config/environments/production.rb", after: ".then { |logger| ::StructuredLogging.new(logger) }") do
```
Comment (sfnelson) [inline template.rb:189]: looks like you've missed staging, unless this happens before the copy from production=>staging

File: template.rb [hunk 2]
```diff
@@ -167,6 +168,41 @@ def setup_timezone
...
+  insert_into_file("config/application.rb", after: "Bundler.require(*Rails.groups)\n") do
+    "require_relative \"../lib/loggers/structured_logging\"\n"
+  end
+
+  gsub_file("config/environments/production.rb", "::Logger::Formatter", "::StructuredLogging::JsonFormatter")
+  gsub_file("config/environments/production.rb", "ActiveSupport::TaggedLogging.new(logger)",
+            "::StructuredLogging.new(logger)")
+
+  insert_into_file("config/environments/production.rb", before: "config.logger") do
+    "STDOUT.sync = true\n  "
```
Comment (sfnelson) [inline template.rb:187]: why is there whitespace after the \n?
you could add the whitespace to the pattern (before) if you're seeing the code inserted in the middle of the line

File: template.rb [hunk 3]
```diff
@@ -167,6 +168,41 @@ def setup_timezone
...
 
+# By default in production Rails is using TaggedLogging
+# replace it with StructuredLogging(with json formatter)
+# and add StructuredLogging(with basic formatter) to development and test
+# Note: should be invoked before #setup_staging
+def setup_logger
+  copy_file("lib/loggers/structured_logging.rb")
+
+  insert_into_file("config/application.rb", after: "Bundler.require(*Rails.groups)\n") do
+    "require_relative \"../lib/loggers/structured_logging\"\n"
```
Comment (sfnelson) [inline template.rb:179]: Rails 7.1 introduced autoloading for lib files. I don't think we should be relying on `require_relative` for lib files unless there's a need to do so, in which case that directory should be added to the list of directories that are excluded from autoloading.

File: lib/loggers/structured_logging.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+require "logger"
+require "active_support/logger"
+
+# Wraps any standard Logger object to provide structured logging capabilities.
+#
+# May be called with a block:
+#
+#   logger = StructuredLogging.new(Logger.new(STDOUT))
+#   logger.context(key: value) { logger.info 'Stuff' } # Logs "{ message: 'Stuff', key: 'value' }"
+module StructuredLogging
```
Comment (sfnelson) [inline lib/loggers/structured_logging.rb:15]: This is really complicated. Could we put it in Koi instead of copying?

**Review by sfnelson — CHANGES_REQUESTED** [review]

File: template.rb (hunk 1 above)
Comment (hasarindaKI) [reply template.rb:189]: Yes, this is done before copying to staging.

File: lib/loggers/structured_logging.rb (hunk 4 above)
Comment (hasarindaKI) [reply lib/loggers/structured_logging.rb:15]: Moved to Koi: https://github.com/katalyst/koi/pull/571

File: template.rb (hunk 3 above)
Comment (hasarindaKI) [reply template.rb:179]: `require "Koi/loggers/structured_logging"`

File: template.rb (hunk 2 above)
Comment (hasarindaKI) [reply template.rb]: This is to align the config block with the correct spacing..

File: template.rb [hunk 5]
```diff
@@ -219,8 +238,7 @@ def install_active_storage
     insert_into_file("config/environments/#{env}.rb", before: /end\Z/) do
       ["",
        "# Store uploaded files on the local file system",
-       "config.active_storage.service = #{service.inspect}\n",
-      ].join("\n  ")
+       "config.active_storage.service = #{service.inspect}\n"].join("\n  ")
```
Comment (sfnelson) [inline template.rb:241]: unnecessary churn

File: template.rb [hunk 6]
```diff
@@ -167,6 +168,24 @@ def setup_timezone
...
+             "STDOUT.sync = true",
+             "config.rails_semantic_logger.add_file_appender = false",
+             "config.semantic_logger.add_appender(io: STDOUT, formatter: :json)"]
+              .join("\n  "))
+
+  gsub_file("config/environments/production.rb", /# Prepend all log lines.*\n/, "")
+  gsub_file("config/environments/production.rb", /config.log_tags.*\n/, "")
+  gsub_file("config/environments/production.rb", /config.logger.*\n/, "")
+  gsub_file("config/environments/production.rb", /\.tap\s* \{ \|logger\|.*\n/, "")
+  gsub_file("config/environments/production.rb", /\.then \{ \|logger\|.*\n/, "")
```
Comment (sfnelson) [inline template.rb:186]: I'd prefer to see this written as a single gsub, both from a safety point of view (don't accidentally remove something else) and from a developer-friendliness point of view. If that's not possible then this is ok, but please add a comment explaining what the source pattern looks like.

**Review by sfnelson — CHANGES_REQUESTED** [review]

**Pull request merged**